File: extra/Records.java
// java --enable-preview --source 15 Records.java

import java.time.*;
import java.util.*;
import java.util.Map.Entry;
import java.util.function.*;
import java.util.stream.*;

import static java.lang.System.out;
import static java.util.Map.entry;

class Records {
    public static void main(String[] args) {
        // exploreRecords();
        exploreMaps();
    }

    static void exploreRecords() {
        var alice = new Employee("Alice", LocalDate.of(2000, 1, 1));
        var alice2 = new Employee(
            alice.name(), alice.hireDate().withYear(2010)
        );
        var alice3 = new Employee(alice2.name(), alice.hireDate());
        out.println(alice);
        out.println(alice.hashCode());
        out.println(alice2.hashCode());
        out.println(alice3.hashCode());
        out.println(alice.equals(alice2));
        out.println(alice.equals(alice3));
    }

    static void exploreMaps() {
        var favorites = new HashMap<>(Map.ofEntries(
            entry("food", "apple"),
            entry("color", "aqua")
        ));
        var favorites2 = new HashMap<>(favorites);
        favorites2.put("food", "avocado");
        // var favorites2 = ((Supplier<Map<String, String>>)() -> {
        //     var result = new HashMap<>(favorites);
        //     result.put("food", "avocado");
        //     return result;
        // }).get();
        var favorites3 = Stream
            .of(favorites2, Map.ofEntries(entry("food", favorites.get("food"))))
            .flatMap(map -> map.entrySet().stream())
            .collect(Collectors.toMap(
                Entry::getKey, Entry::getValue,
                (oldValue, newValue) -> newValue
            ));
        out.println(favorites);
        out.println(favorites2);
        out.println(favorites3);
        out.println(favorites.hashCode());
        out.println(favorites2.hashCode());
        out.println(favorites3.hashCode());
        out.println(favorites.equals(favorites2));
        out.println(favorites.equals(favorites3));
        out.println(favorites == favorites3);
    }
}

record Employee(String name, LocalDate hireDate, Map<String, String> favorites) {
    Employee {
        if (name == null) throw new RuntimeException("No name"); 
        if (hireDate == null) throw new RuntimeException("No hireDate"); 
        if (favorites == null) favorites = Collections.emptyMap();
    }

    Employee(String name, LocalDate hireDate) {
        this(name, hireDate, null);
    }

    public int yearsEmployed() {
        return LocalDate.now().getYear() - hireDate.getYear();
    }
}

record DetailEmployee(Employee employee, Map<String, String> favorites) {}

File: extra/Records.cs
﻿// using System;
// using System.Collections.Generic;
// using System.Collections.Immutable;
// // using System.Text.Json;
// using static System.Console;

// #pragma warning disable CS0162, CS8321
// void LoadTest() {
//     var count = 100000000;
//     if (false) {
//         var structs = new EmployeeStruct[count];
//         for (int i = 0; i < structs.Length; i += 1) {
//             structs[i] = new EmployeeStruct { Name = "Betty" };
//         }
//     } else {
//         var records = new Employee[count];
//         for (int i = 0; i < records.Length; i += 1) {
//             records[i] = new Employee { Name = "Betty" };
//         }
//     }
//     WriteLine("Done!");
//     ReadLine();
// }

// void ExploreMaps() {
//     var favorites = new Dictionary<string, string> {
//         {"color", "aqua"},
//         {"food", "apple"},
//     }.ToImmutableDictionary();
//     var favorites2 = favorites.SetItem("food", "avocado");
//     var favorites3 = favorites2.SetItem("food", "apple");
//     WriteLine(favorites.GetHashCode());
//     WriteLine(favorites2.GetHashCode());
//     WriteLine(favorites3.GetHashCode());
//     WriteLine(favorites == favorites2);
//     WriteLine(favorites == favorites3);
//     // WriteLine(favorites < favorites2);
// }

// void ExploreRecords() {
//     // var alice = new Employee("alice", new (2000, 1, 1));
//     var alice = new Employee {
//         Name = "Alice",
//         HireDate = new (2000, 1, 1),
//         // HireDate = new DateTime(2000, 1, 1),
//     };
//     alice.Validate();
//     WriteLine(alice);
//     WriteLine(alice.YearsEmployed());
//     WriteLine(alice.GetHashCode());

//     var alice2 = alice with { HireDate = new (2010, 1, 1) };
//     var alice3 = alice2 with { HireDate = alice.HireDate };
//     WriteLine(alice2.GetHashCode());
//     WriteLine(alice3.GetHashCode());

//     WriteLine(alice == alice2);
//     // WriteLine(alice < alice2);
//     WriteLine(alice == alice3);

//     var greetings = new Dictionary<Employee, String> {
//         {alice, "Hi!"},
//     }.ToImmutableDictionary();
//     WriteLine(greetings[alice3]);
// }
// #pragma warning restore CS0162, CS8321

// ExploreRecords();
// // ExploreMaps();
// // LoadTest();

// // record Employee(string Name, DateTime HireDate);
// record Employee(
//     string? Name = default,
//     DateTime HireDate = default
//     // Box<DateTime>? HireDate = default
// ) {
//     public string Name { get; init; } = Name!;
//     public DateTime HireDate { get; init; } = HireDate;
//     // public Box<DateTime> HireDate { get; init; } = HireDate!;

//     public void Validate() {
//         if (Name is null) throw new ArgumentException("Name missing");
//         if (HireDate == default) {
//             throw new ArgumentException("HireDate missing");
//         }
//     }

//     public int YearsEmployed() => DateTime.Now.Year - HireDate.Year;
//     // public int YearsEmployed() => DateTime.Now.Year - HireDate.Value.Year;
// }

// record DetailEmployee(
//     string? Name = default,
//     DateTime HireDate = default,
//     IReadOnlyDictionary<string, string>? Favorites = default
// ) : Employee(Name, HireDate) {
//     public IReadOnlyDictionary<string, string>? Favorites { get; init; } =
//         Favorites ?? ImmutableDictionary.Create<string, string>();
// }

// struct EmployeeStruct {
//     public string Name { get; set; }
//     public DateTime HireDate { get; set; }
// }

// record Box<ValueType>(ValueType Value) where ValueType : struct {
//     public static implicit operator Box<ValueType>(ValueType Value)
//         => new Box<ValueType>(Value);

//     public override string ToString() => Value.ToString()!;
// }
